when the path, the class_name or the dn change the resource should be…

NicolasMarcotte opened the pull request:
… recreated

this fix #195

nkatarmal-crest commented:
@NicolasMarcotte Can you please show some testing on your PR by building the provider locally? I am afraid that the forcenew and compute both cannot be true at the same time. I would be happy to merge your request after validating these changes.

NicolasMarcotte commented:
no problem, I will find time to do it this week

NicolasMarcotte commented:
I start from the branch with the following git status
![image](https://user-images.githubusercontent.com/3064154/104933453-ec46ad80-5976-11eb-932d-0fe4818890bb.png)
I changed the makefile locally to install the build so terraform use it.
I clean the build then do a make 
![image](https://user-images.githubusercontent.com/3064154/104933774-519a9e80-5977-11eb-8988-38262fc5ab15.png)
I then go into my test case and install the provider  
![image](https://user-images.githubusercontent.com/3064154/104933997-9cb4b180-5977-11eb-9de7-b5f844c3cc63.png)
I do an apply, here is the plan :
 ![image](https://user-images.githubusercontent.com/3064154/104934143-ca99f600-5977-11eb-946c-35da93c1ac69.png)
and the result
![image](https://user-images.githubusercontent.com/3064154/104934198-d8e81200-5977-11eb-9c24-8c03921180bf.png)
I change the vswitch name so the path and dn change
![image](https://user-images.githubusercontent.com/3064154/104934400-13ea4580-5978-11eb-89eb-317ee61afd13.png)
I should note that it only fix the case where the path or class_name change, the dn attribute being computed, it's value is only know after the apply. I will look into the CustomizeDiffFunc to see if I can fix the case where the dn change but it's path stays constant

NicolasMarcotte commented:
I extracted the method payloadToContainer from the PostAndSetStatus method and 
added a resourceAciRestCustomizeDiff method that compute if the dn has changed

NicolasMarcotte commented:
please tel me if you need more test

nkatarmal-crest commented on aci/resource_rest.go:
```diff
@@ -43,6 +47,7 @@ func resourceAciRest() *schema.Resource {
 			"dn": &schema.Schema{
 				Type:     schema.TypeString,
 				Computed: true,
+				ForceNew: true,
```
Dn is something that should be computed only, we don't want users to pass the Dn. It's never used in the POST/PUT calls. Please remove all the references which checks for the diff in dn and also remove the ForceNew: true from the dn param as well.

NicolasMarcotte replied:
This force new is required for the resource to be recreated when the resourceAciRestCustomizeDiff function tags the dn as changed after apply. I will remove the changes concerning the dn attribute.  However,  this pull request will no longer completely fix the issue as changes to the dn attribute will not recreate the corresponding managed object. I will ping you back when it's done.

lhercot replied:
The DN of an object is always based on one of the fields of the object so if the content is different we should be able to infer that the object need to be recreated. Am I missing something?

NicolasMarcotte replied:
The DN is computed only after the apply, so when Terraform is made aware of the change, it is already to late to recreate the object.

The resourceAciRestCustomizeDiff function computes if the apply will change the dn and mark it as such during the plan phase of the terraform lifecycle.

But if this patch is too involved to review, we (my org) can live a patch that only recreate the object when the path change as the dn can be included in it most situation that we care about.

nkatarmal-crest replied:
My point here is, Dn is not user-configurable, if you pass the dn to this resource in the terraform file we are not going to pass it to the API, the only thing passed to api is content/payload and the path. Dn we calculate based on the object we are going to create. While using Payload mode to create the objects using rest, dn won't be set at all, because we will never know what user tried to do there. There is no guarantee that user will pass the dn in content/payload. So checking the changes in dn is not feasible in this case. If there is something needs to be re-created we might need to think any other way as this resource was never designed to be 100% idempotent based on the nature of content it's going to handle. 


The very simple thing here is you don't need Dn to create an object in the ACI, you need the classname and required attributes, Dn is calculated automatically, So if you change any attribute which requires an object to be recreated checking the changes in Dn won't help as it might not be set in some cases.

NicolasMarcotte replied:
I did the revert and removed the force new on the dn attribute. Thanks for your explanation and sorrry about the delay, we were busy migrating our data-center networks in aci.

nkatarmal-crest commented on aci/resource_rest.go:
```diff
@@ -18,23 +19,26 @@ const ErrDistinguishedNameNotFound = "The Dn is not present in the content"
+		Update:        resourceAciRestUpdate,
+		Read:          resourceAciRestRead,
+		Delete:        resourceAciRestDelete,
+		CustomizeDiff: resourceAciRestCustomizeDiff,
```
I don't think we need this CustomizeDiff function here as we don't need to check for the changes in the Dn. Please remove this function as well.

NicolasMarcotte replied:
I will remove this function when I address the thread on the force new on the dn attribute
